src/blockstore/hashmap_db.zig

```diff
@@ -0,0 +1,254 @@
+    allocator: Allocator,
+    maps: []SharedHashMap,
+    /// shared lock is required to call locking map methods.
+    /// exclusive lock is required to call non-locking map methods.
+    transaction_lock: DefaultRwLock = .{},
+
+    const Self = @This();
+
+    pub const Batch = MapBatch;
+
+    pub fn open(
+        allocator: Allocator,
+        _: Logger,
+        _: []const u8,
+        column_families: []const ColumnFamily,
+    ) !SharedHashMapDB {
+        var maps = try allocator.alloc(SharedHashMap, column_families.len);
+        inline for (0..column_families.len) |i| {
+            maps[i] = try SharedHashMap.init(allocator);
+        }
+        return .{ .allocator = allocator, .maps = maps };
+    }
+
+    pub fn deinit(self_: Self) void {
+        var self = self_;
+        for (self.maps) |*map_| {
+            map_.deinit();
+        }
+        self.allocator.free(self.maps);
+    }
+
+    pub fn put(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
+        key: cf.Key,
+        value: cf.Value,
+    ) !void {
+        const key_bytes = try cf.key().serializeAlloc(self.allocator, key);
+        const val_bytes = try cf.value().serializeAlloc(self.allocator, value);
+        self.transaction_lock.lockShared();
+        defer self.transaction_lock.unlockShared();
+        return try self.maps[cf_index].put(key_bytes, val_bytes);
+    }
+
+    pub fn get(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
+        key: cf.Key,
+    ) !?cf.Value {
+        const key_bytes = try cf.key().serializeAlloc(self.allocator, key);
+        defer self.allocator.free(key_bytes);
+        const map = &self.maps[cf_index];
+
+        self.transaction_lock.lockShared();
+        defer self.transaction_lock.unlockShared();
+        map.lock.lockShared();
+        defer map.lock.unlockShared();
+
+        const val_bytes = map.getPreLocked(key_bytes) orelse return null;
+
+        return try cf.value().deserialize(cf.Value, self.allocator, val_bytes);
+    }
+
+    pub fn getBytes(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
+        key: cf.Key,
+    ) !?BytesRef {
+        const key_bytes = try cf.key().serializeAlloc(self.allocator, key);
+        defer self.allocator.free(key_bytes);
+        var map = self.maps[cf_index];
+
+        self.transaction_lock.lockShared();
+        defer self.transaction_lock.unlockShared();
+        map.lock.lockShared();
+        defer map.lock.unlockShared();
+
+        const val_bytes = map.getPreLocked(key_bytes) orelse return null;
+
+        const ret = try self.allocator.alloc(u8, val_bytes.len);
+        @memcpy(ret, val_bytes);
+        return .{
+            .allocator = self.allocator,
+            .data = ret,
+        };
+    }
+
+    pub fn delete(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
```

Comment:
Same question here, and in general, what's the meaning of these `comptime cf_index`s which appear to only be used at runtime?